R/fn-heatwave.R

```diff
@@ -0,0 +1,393 @@
+    ),
+    measure = dplyr::case_when(
+      element == "Homogenized daily maximum temperature" ~ "daily_max",
+      element == "Homogenized daily minimum temperature" ~ "daily_min",
+      element == "Homogenized daily mean temperature" ~ "daily_mean",
+      TRUE ~ NA_character_
+    ),
+    temp = as.numeric(gsub("[a-zA-Z]", "", temp)),
+    temp = ifelse(temp < -9000, NA_real_, temp)
+  )
+
+  data |>
+    dplyr::select(stn_id, stn_name, measure, date, year = Year,
+                  month = Mo, temp, dplyr::everything(),
+                  -DoM) |>
+    dplyr::filter(!is.na(date)) # Remove invalid dates
+}
+
+write_ahccd_data <- function(zipfiles, save_raw_txt = FALSE,
+                             data_dir = "data") {
+
+  stopifnot(length(zipfiles) == 3L)
+  stopifnot(all(file.exists(zipfiles)))
+
+  fpaths <- future.apply::future_lapply(zipfiles, function(n) {
+    files <- extract_ahccd_data(n, save_raw_txt = save_raw_txt, data_dir = data_dir)
+    files
+  })
+
+  data_dir <- normalizePath(data_dir, mustWork = TRUE)
+
+  parquet_path <- file.path(data_dir, "AHCCD_data", "parquet")
+
+  if (dir.exists(parquet_path)) {
+    del <- unlink(parquet_path, recursive = TRUE, force = TRUE)
+    if (!del) stop("unable to delete existing directory: ", parquet_path)
+  }
+
+  dir.create(parquet_path, recursive = TRUE, showWarnings = FALSE)
+
+  message("Writing parquet files in ", parquet_path, "for ", length(fpaths[[1]]),
+          " stations, partitioning by stn_id, year, and measure")
+
+
+  # this could be done in parallel
+  furrr::future_pwalk(fpaths, function(x, y, z) {
+    d1 <- read_ahccd_data_single(x)
+    d2 <- read_ahccd_data_single(y)
+    d3 <- read_ahccd_data_single(z)
+
+    d <- dplyr::bind_rows(d1, d2, d3)
+    d <- dplyr::group_by(d, stn_id, year, measure)
+    arrow::write_dataset(d, parquet_path, format = "parquet")
+  })
+
+  attr(parquet_path, "timestamp") <- as.character(Sys.time())
+  parquet_path
+}
+
+get_ahccd_stations <- function() {
+  stations_url <- "https://crd-data-donnees-rdc.ec.gc.ca/CDAS/products/EC_data/AHCCD_daily/Homog_Temperature_Stations_Gen3.xls"
+
+  stations_file <- normalizePath(file.path("data", "ahccd_sources", basename(stations_url)),
+                                 mustWork = FALSE)
+  dir.create(dirname(stations_file), showWarnings = FALSE)
+
+  download.file(stations_url, destfile = stations_file, method = "curl")
+
+  colnames <- names(readxl::read_excel(stations_file, skip = 2, n_max = 2))
+
+  stations <- readxl::read_excel(stations_file, skip = 4, col_names = colnames)
+
+  stations <- stations |>
+    janitor::clean_names() |>
+    dplyr::mutate(from = paste(from, x6, sep = "-"),
+                  to = paste(to, x8, sep = "-")) |>
+    dplyr::select(-x6, -x8)
+
+  sf::st_as_sf(stations, coords = c("long_deg", "lat_deg"), crs = 4326)
+}
+
+get_bc_target_stations <- function(stations, buffer, crs) {
+  bc_buff <- sf::st_buffer(st_union(bcmaps::bc_bound()), dist = buffer * 1000)
```

Comment:
tar_make() error:
Error: callr subprocess failed: could not find function "st_union"